### DISCUSSION TITLE
Can sootup do a backwards analysis to construct a call graph?

### DISCUSSION THREAD

**Discussion by dimpdash** [soot-oss/SootUp#1091] [Q&A] [answered]
I am wanting to provide an entry method to find all ways this method is called rather than what it calls. 

Is there a way the entry method can be "destination" of all the calls rather than the source?

**Comment by JonasKlauke** [comment, marked as answer]
There is currently no way to do a backwards analysis. Compared to a forwards analysis you need to know all places in which a call is possible, for CHA it might be possible to develop such an approach, but a higher precision might be difficult. 
A current work around would be to add every method as entry point in the call graph algorithm. That way you would create a complete call graph and then you could extract the subgraph for your "entry method" or you can use the complete call graph for the queries.

**Reply by dimpdash** [reply]
Ah thank you. Appreciate it.

I'm working on a large monorepo, so it's like finding a needle in a haystack. But that workaround should work given time.

**Reply by dimpdash** [reply]
Any thoughts on difficulty of a low precision approach? 

I'm just needing something as good as IntelliJ's caller hierarchy function

**Reply by JonasKlauke** [reply]
I think it would be similar to my workaraound. You have to scan every method body and look for calls to your requested method. The visibility could help to decrease the search area. Private methods are probably only called by the class itself or inner classes. Protected are called by .... Maybe there are already published algorithm to do a backwards analysis, But I dont know them.
